change(ResultList) Use ID and ClassName for objects

nyeholt opened the pull request:
Rather than indexing and mapping the PHP class Type and ID to
_type and _id, swap to using explicit fields ID and ClassName. The _id and
_type fields are deprecated in 6.0, meaning potential collisions in the _id
field if just left as the SS ID. Instead, the _id field is changed
to being a concatenation of something like

`$id = $this->owner->getElasticaType() . '_' . $this->owner->ID;`

The extension to this is that it allows me to subsequently alter the $document->id to include the $stage value also, so that further updates / deletes are done across appropriate stages too. 


https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-type-field.html

nyeholt committed:
change(ResultList) Use ID and ClassName for objects

Rather than indexing and mapping the PHP class Type and ID to
_type and _id, swap to using explicit fields ID and ClassName. The _id and
_type fields are deprecated in 6.0, meaning potential collisions in the _id
field if just left as the SS ID. Instead, the _id field is changed
to being a concatenation of something like

```
$id = $this->owner->getElasticaType() . '_' . $this->owner->ID;
```

https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-type-field.html

stecman commented on src/ResultList.php:
```diff
@@ -171,10 +176,17 @@ public function toArray()
+                    //      Swapped to using the ClassName and ID fields introduced elsewhere
+                    //      due to _type being deprecated
+                    //      https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-type-field.html
+                    $type = isset($data['ClassName']) ? $data['ClassName'] : $item->getType();
```
Minor nit - no need to name and date comments as that info is accessible with `git blame`

nyeholt replied:
Yep - it's become an internal practice just to avoid needing to go back in time to get that context :)

stecman commented:
Otherwise seems ok.

@shane-garelja, by chance is anyone available to test this change out against an existing project? I don't have any projects that use this module, so can't do this quickly.

nyeholt committed:
fix(ResultList) Use the correct  variable when adding to existing  list

nyeholt commented:
Just a quick bump here - we're using a forked version of the module as we need this functionality - any chance of it making it back?

silbinarywolf commented:
@stecman or @shane-garelja Can you please make some time to review / test this and accept the PR?

clairefinnie commented:
hey @nyeholt  & @silbinarywolf, I will make some time to review this, hopefully by tomorrow.

silbinarywolf commented:
Awesome! Thanks a lot!

shane-nzer commented:
@bendubuisson or @clairefinnie - this seems to have dragged on a bit. Do either of you have projects you can quickly test this on?

silbinarywolf commented:
@shane-garelja , @bendubuisson  or @clairefinnie - Can one of you please make time for this?

nyeholt commented:
Just another bump on this, as we're still using a forked version for now.

nyeholt committed:
refactor(ResultList) handle existing 'source'

Changed setting of '_source' param to account for users binding one into
place prior to querying.

nyeholt committed:
fix(ResultList) use hasParam to ensure _source exists first

tardinha commented:
Can we get this pulled in???
